pages/DSNP/Graph.md

```diff
@@ -2,29 +2,126 @@
+    Care should be taken to ensure that User Id values greater than or equal to 2<sup>63</sup>, where used by an implementation, are converted correctly between signed and unsigned representations.</mark>
+* Timestamp when relationship was created, in seconds since the [Unix epoch](https://en.wikipedia.org/wiki/Unix_time)
 
-Implementation coming in the future. See the [DSNP Whitepaper](https://github.com/LibertyDSNP/papers) for foundational ideas.
+GraphEdge objects MUST conform to the following [Avro](https://avro.apache.org) schema:
 
-### Following
+```
+{
+    "namespace": "org.dsnp.userdata",
+    "name": "GraphEdge",
+    "type": "record",
+    "doc": "A relationship to another DSNP user",
+    "fields": [
+        {
+            "name": "userId",
+            "type": "long",
+            "doc": "DSNP User Id of object of relationship"
+        },
+        {
+            "name": "since",
+            "type": "long",
+            "doc": "Timestamp in Unix epoch seconds when this relationship was originally established"
+        }
+    ]
+}
+```
 
-A "follow" is the act of publicly following a [DSNP User Identity](Identifiers.md#dsnp-user-id), which results in adding this DSNP User Id to a user's social graph.
+## User Data Sets
 
-An "unfollow" is the act of publicly unfollowing a [DSNP User Identity](Identifiers.md#dsnp-user-id) which results in the removal of this DSNP User Id from a user's social graph.
+DSNP user data structures capture three distinct sets of relationships, as described below.
 
-### Friendship
+### Public Follows
 
-There is currently no concept of "friendship" within the DSNP network.
-Friendship requires a mutual acknowledgment between two different DSNP User Identities.
-Friendship could be thought of as "mutual following" - where two DSNP User Identities are following each other.
+Each public follow records a unidirectional relationship where one user (the controller of the graph) is interested in following the public speech of another user.
+When follows are public, they are visible to any other user or system, including the user being followed.
+
+The `publicFollows` user data type is composed of a set of GraphEdge objects, which is then compressed.
+
+### Private Follows
+
+A private follow also records an interest in following the public speech of another user, but the data is stored in encrypted form, meaning the relationship is not visible to the followed user, or any other user, without access to the decryption key.
+
+The `privateFollows` user data type is composed of a set of GraphEdge objects, which is then compressed and encrypted.
+
+### Private Connections
+
+Private connections represent one half of a bidirectional relationship between two DSNP Users.
+An encrypted data set controlled by a DSNP user records which users they recognize as connections.
+A second, public, data set enables users to signal their connection to one another via Pseudonymous Relationship Identifiers.
```

Comment:
The mention here of "second" feels like it references the statement about "half" above it.

Perhaps a rewording like this?:

```suggestion
Private connections are stored in on both sides of the bidirectional relationship between two DSNP Users.
Each user can then update the status of the relationship without loss of privacy.
Each DSNP user maintains and controls two sets of data:
1. An encrypted data set that records which users are recognized as connections.
2. A public data set that is to signal connection status via Pseudonymous Relationship Identifiers.
```